src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -209,21 +212,31 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
             if (idToken != null)
             {
                 msalIdTokenCacheItem = new MsalIdTokenCacheItem
-                    (preferredEnvironmentHost, requestParams.ClientId, response, tenantId);
+                    (preferredEnvironmentHost, requestParams.ClientId, response, tenantId, subject);
             }
 
             lock (LockObject)
             {
                 try
                 {
+                    Account account = null;
+                    if (msalAccessTokenCacheItem.HomeAccountId != null)
+                    {
+                        account = msalAccessTokenCacheItem.IsAdfs
+                                      ? new Account(
+                                          msalAccessTokenCacheItem.HomeAccountId,
+                                          idToken?.Upn,
+                                          msalAccessTokenCacheItem.Environment)
+                                      : new Account(
+                                          msalAccessTokenCacheItem.HomeAccountId,
+                                          preferredUsername,
+                                          preferredEnvironmentHost);
+                    }
```

Comment:
this looks to be the same code as in AutenticationResult. Perhaps this should be a factory for creating the Account object?

Reply:
Not sure a factory method would be ideal since it'd still be kind of ugly/take 5 params. What about just setting a username/environemnt variable based on whether its adfs or not? And then 1 constuctor call would be sufficienitn

Reply:
what makes me wonder is that here you are creating an account with home account id and other params set similar to regular accounts. however in the authentication results you set it using 1 parameter. I believe the special account contructs should be done in the account object. to me that is one of the things factory helps doing (besides from encapsulating object construction)

---
In reply to: [252398617](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/834#discussion_r252398617) [](ancestors = 252398617)

Reply:
however it may in this case be more obvious to stick to what you have. 

---
In reply to: [255316521](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/834#discussion_r255316521) [](ancestors = 255316521,252398617)